wfe2/wfe.go

```diff
@@ -1486,21 +1552,154 @@ func (wfe *WebFrontEndImpl) Order(ctx context.Context, logEvent *web.RequestEven
-		Error:          string(order.Error),
+	switch request.Method {
+	case "GET", "HEAD":
+		wfe.getOrder(response, request, order, logEvent)
+		return
+	case "POST":
+		wfe.finalizeOrder(ctx, response, request, order, logEvent)
+		return
+	default:
+		// We use ServerInternal here because this handler was only registered for
+		// GET, HEAD and POST. Anything else making it this far is an internal
+		// error!
+		wfe.sendError(response, logEvent,
+			probs.ServerInternal(fmt.Sprintf(
+				"Unsupported HTTP method %q", request.Method)), nil)
+		return
 	}
-	for i, authzID := range order.Authorizations {
-		respObj.Authorizations[i] = wfe.relativeEndpoint(request, fmt.Sprintf("%s%s", authzPath, authzID))
+}
+
+func (wfe *WebFrontEndImpl) getOrder(
+	response http.ResponseWriter,
+	request *http.Request,
+	order *corepb.Order,
+	logEvent *web.RequestEvent) {
+	respObj := wfe.orderToOrderJSON(request, order)
+	err := wfe.writeJsonResponse(response, logEvent, http.StatusOK, respObj)
+	if err != nil {
+		wfe.sendError(response, logEvent, probs.ServerInternal("Error marshaling order"), err)
+		return
+	}
+}
+
+// finalizeOrder accepts a POST request to the order endpoint intended to
+// finalize an order by invoking the issuance process to produce a certificate
+// for the order. Most processing of the order details is handled by the RA but
+// we do attempt to throw away requests with invalid CSRs or made by accounts
+// that have not agreed to the terms of service here.
+func (wfe *WebFrontEndImpl) finalizeOrder(
+	ctx context.Context,
+	response http.ResponseWriter,
+	request *http.Request,
+	order *corepb.Order,
+	logEvent *web.RequestEvent) {
+	// Validate the POST body signature and get the authenticated account for this
+	// finalize order request
+	body, _, acct, prob := wfe.validPOSTForAccount(request, ctx, logEvent)
+	addRequesterHeader(response, logEvent.Requester)
+	if prob != nil {
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+
+	// If the authenticated account ID doesn't match the order's registration ID
+	// pretend it doesn't exist and abort.
+	if acct.ID != *order.RegistrationID {
+		wfe.sendError(response, logEvent, probs.NotFound(fmt.Sprintf("No order found for account ID %d", acct.ID)), nil)
+		return
+	}
+
+	// The account must have agreed to the subscriber agreement to finalize an
+	// order since it will result in the issuance of a certificate.
+	// Any version of the agreement is acceptable here. Version match is enforced in
+	// wfe.Registration when agreeing the first time. Agreement updates happen
+	// by mailing subscribers and don't require a registration update.
+	if acct.Agreement == "" {
+		wfe.sendError(response, logEvent,
+			probs.Unauthorized("Must agree to subscriber agreement before any further actions"), nil)
+		return
+	}
+
+	// If there is already a certificate serial for this order then it has been
```

Comment:
This comment doesn't seem to match the check.

Reply:
Oops. This was leftover from before I remembered the Order status field existed and was determining if an order was finalized or not strictly based on the presence of the certificate serial.